### PR TITLE
Added simple disjoint path computation

### PR DESCRIPTION
**Pull request by NvanAdrichem**
This is the first part of some work I did on disjoint path computation. I have another set in which I implemented the nearly-full Bhandari (including node-splitting, k>2 disjoint paths, finding reverse overlapping paths, etc.) for which I will open another PR.

Please let me know your first feedback on this implementation of finding disjoint paths through iteratively removing found paths, then I can rewrite accordingly and also write up the right documentation. I am very much open to another function name, since I don't feel it is straight enough about the downsides of iteratively removing paths (no optimizations besides first shortest path and susceptible to trap topologies), neither can I find a generalized common name nor who was the first to propose this method.

P.s. I think the check on negative weights could also be put into the generic shortest path functions. Also, they cannot return both path and length in one function call which implies redundant computation when you want both and want to use the generic functions. If you want to I could change those in a separate PR, but we should maybe discuss the conventions of the function names in the generic and unweighted shortest path functions on beforehand.

### CODE REVIEW

**Comment by NvanAdrichem** [comment]
I may need some help on the checks failing, I think I put all imports on the right place. Nosetests also passes without problems.

File: networkx/algorithms/__init__.py [hunk 1]
```diff
@@ -46,6 +47,7 @@
 import networkx.algorithms.components
 import networkx.algorithms.connectivity
 import networkx.algorithms.coloring
+import networkx.algorithms.disjoint_paths
```
Comment (Michael-E-Rose) [inline networkx/algorithms/__init__.py]: Why do you want to import `networkx.algorithms.disjoint_paths` twice?

File: networkx/algorithms/disjoint_paths/simple_disjoint.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+"""
+
+#    Copyright (C) 2015-2016 by
+#    Niels L. M. van Adrichem <n.l.m.vanadrichem@tudelft.nl>
+#    All rights reserved.
+#    BSD license.
+
+__author__ = """\n""".join(['Niels L. M. van Adrichem <n.l.m.vanadrichem@tudelft.nl>'])
+__all__ = ['simple_disjoint']
+
```
Comment (Michael-E-Rose) [inline networkx/algorithms/disjoint_paths/simple_disjoint.py]: The head is not precisely as it should be according to networkx standards. Please see https://github.com/networkx/networkx/wiki#user-content-top-of-the-module

File: networkx/algorithms/disjoint_paths/simple_disjoint.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+def simple_disjoint(G, source, target, weight=None, k=2, node_disjoint=False):
+    """ Simplest, even naive, disjoint path computation by removing previously
+    found paths and computing the next shortest path. This method, although it 
+    often works out, gives no guaranteed correct result since it is susceptibl
+    e to trap topologies where the removal of the shortest path blocks finding
+    2 non-shortest path disjoint paths. Furthermore, besides the first
+    found path equalling the shortest path, it doesn't optimize for path
+    lengths.
+
```
Comment (Michael-E-Rose) [inline networkx/algorithms/disjoint_paths/simple_disjoint.py]: The docstring needs improvement, e.g. list of parameters, examples, etc. The first sentence should summarize the rest and be followed by a more detailed explanation, with a blank line between both. See https://github.com/networkx/networkx/wiki#user-content-docstring-formatting

Also, there should be two blanks between two sentences.

File: networkx/algorithms/__init__.py [hunk 4]
```diff
@@ -11,6 +11,7 @@
 from networkx.algorithms.cycles import *
 from networkx.algorithms.cuts import *
 from networkx.algorithms.dag import *
+from networkx.algorithms.disjoint_paths import *
```
Comment (Michael-E-Rose) [inline networkx/algorithms/__init__.py]: The problem with your importing is this line. You try to import a file in folder `.\networkx\tree\master\networkx\algorithms`, but you want to import `__init.py__` in folder `.\networkx\tree\master\networkx\algorithms`, which is what the line below does.

**Comment by jtorrents** [comment]
Hi @NvanAdrichem , coping the Graph and then removing the shortest paths found is very slow, especially in the case that you need to find node or edge disjoint paths among several, potentially many, pairs of nodes in the same graph (is that your use case?).

This function is very general as it deals with both weighted and unweighed cases (and both for node and edge disjoint paths if I understand correctly). But we already have functions that can deal a lot faster with some of this cases. The unweighed node and edge disjoint path problem can be formulated in terms of connectivity as the local node connectivity of a pair of nodes of a graph is equal to the number of node disjoint paths between (same for edges and edge connectivity). Right now there is no nice API to get the disjoint paths but they are computed in the connectivity package.

For the case of the approximation to node disjoint paths removing the paths already found, in terms of connectivity, we have an implementation of White and Newman approximation algorithm to node connectivity that does exactly this. That implementation however, uses a custom version of the `bidirectional_shortest_path` function that accepts a container of nodes to ignore during the BFSs search. Again, the node disjoint paths are computed but there is no nice API to get them. See:

https://github.com/networkx/networkx/blob/master/networkx/algorithms/approximation/connectivity.py#L107 

In summary, I think we should think first about a nice API for node and edge disjoint paths and implement it reusing what we already have or following the same approach (ie custom BFS instead of copying/removing, or even better a Graph view? in the approximation case and flow algorithms in the exact case for unweighed graphs). Regarding the API I think that node and edge disjoint paths should be separate functions. The exact (flow based) ones should be in the global namespace (ie `nx.node_disjoint_paths`) and the approximation ones in the approximation package, so users will have to import them explicitly, following the pattern of `nx.node_connectivity`.

File: networkx/algorithms/__init__.py (hunk 4 above)
Comment (NvanAdrichem) [inline networkx/algorithms/__init__.py]: Thanks for your comments, I have updated the header. The docstring and docs is something I will fix once the code is finished. I think it should be imported like this, in order to move the function names in disjoint_paths to the top of the namespace (such as nx.function_name()) instead of having to write out the complete module path. Furthermore, it looks identical to how most other modules (such as shortest_paths and assortativity) are imported. Is there perhaps some additional settings for the test suites I miss to include a new module? If I remove this line the function gets lost from the namespace and nosetests fails.

Comment (NvanAdrichem) [inline networkx/algorithms/__init__.py]: Nevermind, I just found it. I browsed a bit more carefully through the build log and saw it performed an actual setup of networkx. I had to add the new module name to setup.py...

**Comment by NvanAdrichem** [comment]
Hi @jtorrents, thank you for thinking along. I had not considered looking into connectivity at all (I'm in computer networks, not graph theory, so it does not always come natural), so I had not seen that implementation of bidirectional shortest path.

To me, one function that can compute both weighted, unweighted and node- and edge-disjoint paths feels practically intuitive. I agree that copying and modifying the graph is very slow, so to solve that could be to extend the shortest path computation to exclude certain nodes and edges like in approximation.connectivity. I would prefer to extend the general shortest path implementation to include the exclude parameter to avoid redundant code. However, I also have work on Bhandari's algorithm that I try to open a PR for this weekend, which does even more graph modifications, though relies on the Bellman-Ford algorithm (due to the insertion of negative link weights). This would imply that any function that needs such a shallow overlay needs to be adapted. For overall adaptability (also to other functions) I opt to implement the overlay graph class you suggest which represents a view of an existing ([Multi][Di])Graph, only stores changes by overriding the __getattr__ and __setattr__ functions, and prohibits access to mutable objects in the dictionaries of the original objects other than through a "get-copy-of" function to prevent unauthorized modification of the original containers.

I would like to include that, but as a separate PR though. I would prefer this PR to be about correctness of the algorithm and agreement on the naming and API, after which we can develop the overlay graph class and adapt the functions that could use such a graph (such as this function, the Bhandari function I am wrapping up, the approximation.connectivity function, and even more).

**Comment by jtorrents** [comment]
@NvanAdrichem I've added a flow based implementation of node and edge disjoint paths in #2063 . 

Regarding the API, a function that returns different things (eg node or edge disjoint paths) depending on a flag as argument requires the user to look up the documentation more often. It is better self explanatory function names that always return the same kind of thing.

**Comment by jtorrents** [comment]
Regarding "weighted" node and edge disjoint paths, I think that instead of a general concept it is more a measure on which paths to select in this approximation approach to disjoint paths. That is, given that you can fall in these "topological traps" you select the paths with less weight. But this does not improve the approximation as you still miss other paths. Node or edge disjoint paths are so regardless of their weight, and we know thanks to graph theory that there are as many {node|edge} disjoint paths as the value of {node|edge} connectivity.

For example, the graph that you use in the test for this PR:

figure_1

This approximation finds only two node or edge disjoint paths between `A` and `Z` whether you use plain bidirectional BFS or weighted algorithims, the only thing that changes are the paths that you find, not their number. But there are more. With the flow based implementation from #2063 :

```python
In [115]: %paste
    G = nx.Graph()
    G.add_edges_from([
        ('A', 'B'), ('A', 'D'), ('A', 'F'), ('A', 'G'),
        ('B', 'C'), ('B', 'D'), ('B', 'G'), ('C', 'D'),
        ('C', 'E'), ('C', 'Z'), ('D', 'E'), ('D', 'F'),
        ('E', 'F'), ('E', 'Z'), ('F', 'Z'), ('G', 'Z')])
## -- End pasted text --

In [116]: list(nx.edge_disjoint_paths(G, 'A', 'Z'))
Out[116]: [['A', 'B', 'C', 'Z'], ['A', 'D', 'E', 'Z'], ['A', 'G', 'Z'], ['A', 'F', 'Z']]

In [117]: list(nx.node_disjoint_paths(G, 'A', 'Z'))
Out[117]: [['A', 'F', 'Z'], ['A', 'G', 'Z'], ['A', 'B', 'C', 'Z'], ['A', 'D', 'E', 'Z']]

In [118]: nx.edge_connectivity(G, 'A', 'Z')
Out[118]: 4

In [119]: nx.node_connectivity(G, 'A', 'Z')
Out[119]: 4
```

**Comment by jtorrents** [comment]
Regarding shortest paths computations excluding nodes and/or edges, see #762 for discussion on different approaches. The last comment on that issue from @jfinkels seems specially relevant.

**Comment by NvanAdrichem** [comment]
That's a pretty cool system to exclude and even temporarily change link weights without altering the actual graph.

Initially I was more thinking to create a suite that contained multiple shortest paths algorithms, just like the shortest_paths.weighted module besides Dijkstra's also contains Bellman-Ford, Johnson's, etc. Of which this would be the first, then add Bhandari's, per-hop single-node and single-edge disjointness, and perhaps more like Suurballe-Tarjan, etc. I did not intend this to be the default edge- or node-disjoint functions in NetworkX, I wanted to create a sub-module from which one could choose which one to use to its own benefits (single-source, all-to-all, etc.) Considering you have already implemented a better solution than this implementation I suggest we close this PR.

Unless you think there is more support for a module containing multiple disjoint-path approaches, I will seize those activities.

**Comment by jtorrents** [comment]
@NeilGirdhar didn't know about Bhandari's and Suurballe algorithms. After a quick read they seem like a good match for NetworkX (and now I see why you only get 2 disjoint paths in your implementation ;) ), So, please if you have time and energy keep working on it. Also the approaches to make this kind of restricted shortest paths algorithms fast (such as edge weight functions, graph views, ...) can help to speed up other parts of NetworkX if we find a nice and consistent way to filter nodes and edges in standard algorithms such as shortest paths.

The flow based implementations of disjoint paths are a baseline (ie the classical way to solve this kind of problem) I think that all other algorithms that solve similar problems *and* are faster than the flow solution are a good match for NetworkX (by faster I mean more than twice as fast). I posted some benchmarks at #2063  and the approximation to node disjoint paths from `approximation.node_connectivity` is about ten times faster than the faster flow algorithm that we have. I plan to add a function `approximation.node_disjoint_paths`.
